React/NextJs/ProjectStructure.md

```diff
@@ -0,0 +1,212 @@
+        └── index.tsx
+```
+
+With atomic design, we would completely separate all UI components and build them as a blocks of atoms, molecules and organisms (together with templates and pages for putting everything together at the end) 
+
+```
+src
+.
+└── components
+    └── ui
+        ├── atoms
+        │   └── Button
+        │       ├── Button.tsx
+        │       └── stories.tsx
+        ├── molecules
+        │   ├── Card
+        │   │   └── index.tsx
+        │   └── WelcomeBanner
+        │       └── index.tsx
+        └── organisms
+            └── Carousel
+                └── index.tsx
+```
+
+## Different molecule layouts
+
+Sometimes, you might want to create a molecule specific for mobile and desktop. You might use a tool like [fresnel](https://github.com/artsy/fresnel) for detecting mediaqueries.
+
+In this case, inside a specific molecule, we could add a subfolder `layouts` (not to be confused with actual layout described below) to define how our molecule would look like on specific media query.
+
+```
+.
+└── molecules
+    └── Card
+        ├── layouts
+        │   ├── Desktop.tsx
+        │   └── Mobile.tsx
+        └── index.tsx
+```
+
+For this case, inside `index.tsx` we would have something like this:
+```tsx
+  export const Card = () => {
+    return (
+      <Media at="sm">
+        <Mobile />
+      </Media>
+      <Media at="md">
+        <Desktop />
+      </Media>
+    )
+  }
+```
+
+This is the case **only** for layouts that would add too much complexity when using standard css queries.
+
+
+### Layouts
+
+With Layouts, we generally define overall page structure that will be presented to the user based on a specific auth state or a specific page.
+
+For example, your app could have an Admin dashboard on `/admin` route which has a fixed header and sidebars on both right and left, and scrollable content in the middle (you're most likely familiar with this kind of layout).
+
+You will define your different layouts inside `layouts` folder:
+
+```
+.
+.
+└── components
+    └── layouts
+        ├── AdminLayout
+        │   └── index.tsx
+        ├── MainLayout
+        │   └── index.tsx
+        └── BlogLayout
+            └── index.tsx
+```
+
+Then, when creating your routes (pages), you will wrap your page in the layout that represents the current page:
+
+```tsx
+// pages/index.tsx
+
+export default function Home() {
+  return (
+    <MainLayout>
+      ...
+    </MainLayout>
+  )
+}
+
+// pages/admin.tsx
+export default function Admin() {
+  return (
+    <AdminLayout>
+      ...
+    </AdminLayout>
+  )
+}
+```
+
+## Setting up store
+
+Your datx store will be placed in the root of the `src` folder as follows:
+
+```
+src
+└── store
+    ├── models
+    │   ├── User.ts
+    │   └── Session.ts
+    ├── utlis
+    │   ├── config.ts
+    │   └── initializeStore.ts
+    └── index.ts
+```
+
+## Fetchers
+
+In `src/fetchers` you will organize your (swr) fetchers, usually by the model on which you will make API calls.
```

Comment:
Not a big deal, but a bit strange folder hierarchy with models and fetchers.

Models are in `src/store/models` and fetchers that are in  `src/fetchers` are using those models. I would expect that the models and fetchers dirs are siblings (on the same level) instead of models being in a subdir.

Would one of the following options make sense?

1.
```
src/
  - store/
  - models/
  - fetchers/
```
2.
```
src/
  - store/
    - models/
    - fetchers/
```

Reply:
I wouldn't mind if the models were in src - they are quite important and nesting them inside of the store folder is basically just mirroring a datx implementation detail, which is not necessary.

Reply:
Fetchers might not be related to the store (for example, if you are using `swr` to fetch google places), so I like the first, simplest option better. Also, it always bothered me that models are inside the store folder, although they are used globally (their purpose is not only for using them with the store, but also to get autocomplete inside components, utils, etc.).